### PR TITLE
Minimize failover impact

### PR DESCRIPTION
**Pull request by wxing1292**
* Add metrics to domain cache
* Use separate max tps for failover
* Use service specific (frontend, matching, history) host level persistence layer throttling.

Note:
Still tweaking on the throttle limit

### CODE REVIEW

File: common/persistence/persistenceMetricClients.go [hunk 1]
```diff
@@ -379,6 +379,8 @@ func (p *workflowExecutionPersistenceClient) updateErrorMetric(scope int, err er
 	case *TimeoutError:
 		p.metricClient.IncCounter(scope, metrics.PersistenceErrTimeoutCounter)
 		p.metricClient.IncCounter(scope, metrics.PersistenceFailures)
+	case *workflow.LimitExceededError:
```
Comment (samarabbas) [inline common/persistence/persistenceMetricClients.go]: LimitExceededError is application level error.  I don't think we return this error anywhere in the persistence code.  Use ServerBusy for any throttling.

File: common/persistence/persistenceMetricClients.go [hunk 2]
```diff
@@ -72,53 +72,53 @@ var _ HistoryManager = (*historyPersistenceClient)(nil)
 var _ MetadataManager = (*metadataPersistenceClient)(nil)
 var _ VisibilityManager = (*visibilityPersistenceClient)(nil)
 
-// NewShardPersistenceClient creates a client to manage shards
-func NewShardPersistenceClient(persistence ShardManager, metricClient metrics.Client, logger bark.Logger) ShardManager {
+// NewShardMetricsPersistenceClient creates a client to manage shards
+func NewShardMetricsPersistenceClient(persistence ShardManager, metricClient metrics.Client, logger bark.Logger) ShardManager {
```
Comment (samarabbas) [inline common/persistence/persistenceMetricClients.go]: rename to ShardPersistenceMetricsClient.

File: common/persistence/persistenceMetricClients.go [hunk 3]
```diff
@@ -72,53 +72,53 @@ var _ HistoryManager = (*historyPersistenceClient)(nil)
...
 		persistence:  persistence,
 		metricClient: metricClient,
 		logger:       logger,
 	}
 }
 
-// NewWorkflowExecutionPersistenceClient creates a client to manage executions
-func NewWorkflowExecutionPersistenceClient(persistence ExecutionManager, metricClient metrics.Client, logger bark.Logger) ExecutionManager {
+// NewWorkflowExecutionMetricsPersistenceClient creates a client to manage executions
+func NewWorkflowExecutionMetricsPersistenceClient(persistence ExecutionManager, metricClient metrics.Client, logger bark.Logger) ExecutionManager {
```
Comment (samarabbas) [inline common/persistence/persistenceMetricClients.go]: rename to WorkflowExecutionPersistenceMetricsClient.

File: common/persistence/persistenceMetricClients.go [hunk 4]
```diff
@@ -72,53 +72,53 @@ var _ HistoryManager = (*historyPersistenceClient)(nil)
...
 		persistence:  persistence,
 		metricClient: metricClient,
 		logger:       logger,
 	}
 }
 
-// NewTaskPersistenceClient creates a client to manage tasks
-func NewTaskPersistenceClient(persistence TaskManager, metricClient metrics.Client, logger bark.Logger) TaskManager {
+// NewTaskMetricsPersistenceClient creates a client to manage tasks
+func NewTaskMetricsPersistenceClient(persistence TaskManager, metricClient metrics.Client, logger bark.Logger) TaskManager {
```
Comment (samarabbas) [inline common/persistence/persistenceMetricClients.go]: rename to TaskPersistenceMetricsClient.

File: common/persistence/persistenceMetricClients.go [hunk 5]
```diff
@@ -72,53 +72,53 @@ var _ HistoryManager = (*historyPersistenceClient)(nil)
...
 		persistence:  persistence,
 		metricClient: metricClient,
 		logger:       logger,
 	}
 }
 
-// NewHistoryPersistenceClient creates a HistoryManager client to manage workflow execution history
-func NewHistoryPersistenceClient(persistence HistoryManager, metricClient metrics.Client, logger bark.Logger) HistoryManager {
+// NewHistoryMetricsPersistenceClient creates a HistoryManager client to manage workflow execution history
+func NewHistoryMetricsPersistenceClient(persistence HistoryManager, metricClient metrics.Client, logger bark.Logger) HistoryManager {
```
Comment (samarabbas) [inline common/persistence/persistenceMetricClients.go]: rename to NewHistoryPersistenceMetricsClient.

File: common/persistence/persistenceMetricClients.go [hunk 6]
```diff
@@ -72,53 +72,53 @@ var _ HistoryManager = (*historyPersistenceClient)(nil)
...
 		persistence:  persistence,
 		metricClient: metricClient,
 		logger:       logger,
 	}
 }
 
-// NewMetadataPersistenceClient creates a MetadataManager client to manage metadata
-func NewMetadataPersistenceClient(persistence MetadataManager, metricClient metrics.Client, logger bark.Logger) MetadataManager {
+// NewMetadataMetricsPersistenceClient creates a MetadataManager client to manage metadata
+func NewMetadataMetricsPersistenceClient(persistence MetadataManager, metricClient metrics.Client, logger bark.Logger) MetadataManager {
```
Comment (samarabbas) [inline common/persistence/persistenceMetricClients.go]: rename to NewMetadataPersistenceMetricsClient.

File: common/persistence/persistenceMetricClients.go [hunk 7]
```diff
@@ -72,53 +72,53 @@ var _ HistoryManager = (*historyPersistenceClient)(nil)
...
 		persistence:  persistence,
 		metricClient: metricClient,
 		logger:       logger,
 	}
 }
 
-// NewVisibilityPersistenceClient creates a client to manage visibility
-func NewVisibilityPersistenceClient(persistence VisibilityManager, metricClient metrics.Client, logger bark.Logger) VisibilityManager {
+// NewVisibilityMetricsPersistenceClient creates a client to manage visibility
+func NewVisibilityMetricsPersistenceClient(persistence VisibilityManager, metricClient metrics.Client, logger bark.Logger) VisibilityManager {
```
Comment (samarabbas) [inline common/persistence/persistenceMetricClients.go]: rename to NewVisibilityPersistenceMetricsClient.

File: common/persistence/persistenceMetricClients.go (hunk 1 above)
Comment (wxing1292) [reply common/persistence/persistenceMetricClients.go]: sure

**Pull request merged**